File: src/services/odsaCommon.ts
/**
 * ODSA Common Types and Helpers
 *
 * Shared interface for configData JSONB used by ap2006/ap2009
 * and helper to build ODSA response configs.
 */

import type { ApplicationConfig } from '../protocol/responseTypes.js';

export interface OdsaConfigData {
  subscriptionState?: 'eligible' | 'pending' | 'active' | 'suspended';
  smdpAddress?: string;
  profileIccid?: string;
  profileType?: string;
  companionDeviceImei?: string;
  primaryDeviceImsi?: string;
  planId?: string;
  planName?: string;
  serviceFlowUrl?: string;
  serviceFlowUserData?: string;
}

export interface OdsaContext {
  operation?: string;
  operationType?: number;
}

/**
 * Build a base ODSA ApplicationConfig with SubscriptionResult in extraParams.
 */
export function buildOdsaBaseConfig(
  appId: string,
  status: number,
  provStatus: number,
  tcStatus: number,
  subscriptionResult: number,
  extraFields?: Record<string, string>,
): ApplicationConfig {
  const extraParams: Record<string, string> = {
    SubscriptionResult: String(subscriptionResult),
    ...extraFields,
  };

  return {
    appId,
    entitlementStatus: status,
    provStatus,
    tcStatus,
    extraParams,
  };
}

File: internal/services/odsa_common.go
package services

import (
	"strconv"

	"github.com/hkassaei/poc-entitlements-as-a-service-go/internal/protocol"
)

// OdsaConfigData holds the configData JSONB for ODSA services (ap2006/ap2009).
type OdsaConfigData struct {
	SubscriptionState   string `json:"subscriptionState,omitempty"`
	SMDPAddress         string `json:"smdpAddress,omitempty"`
	ProfileICCID        string `json:"profileIccid,omitempty"`
	ProfileType         string `json:"profileType,omitempty"`
	CompanionDeviceIMEI string `json:"companionDeviceImei,omitempty"`
	PrimaryDeviceIMSI   string `json:"primaryDeviceImsi,omitempty"`
	PlanID              string `json:"planId,omitempty"`
	PlanName            string `json:"planName,omitempty"`
	ServiceFlowURL      string `json:"serviceFlowUrl,omitempty"`
	ServiceFlowUserData string `json:"serviceFlowUserData,omitempty"`
}

// OdsaContext holds the operation context from the request.
type OdsaContext struct {
	Operation     string
	OperationType int
}

// BuildOdsaBaseConfig builds a base ODSA ApplicationConfig with SubscriptionResult in extraParams.
func BuildOdsaBaseConfig(appID string, status, provStatus, tcStatus, subscriptionResult int, extraFields map[string]string) *protocol.ApplicationConfig {
	extra := map[string]string{
		"SubscriptionResult": strconv.Itoa(subscriptionResult),
	}
	for k, v := range extraFields {
		extra[k] = v
	}

	return &protocol.ApplicationConfig{
		AppID:             appID,
		EntitlementStatus: status,
		ProvStatus:        protocol.IntPtr(provStatus),
		TcStatus:          protocol.IntPtr(tcStatus),
		ExtraParams:       extra,
	}
}
